systemtest/benchtest/tracehandler/handler.go

```diff
@@ -0,0 +1,216 @@
+	// items contains the number of events (minus metadata) in the batch.
+	items uint
+}
+
+type Handler struct {
+	transport    transport.Transport
+	batches      []batch
+	warmupEvents uint
+}
+
+// New creates a new tracehandler.Handler
+func New(p string, t transport.Transport, storage fs.FS, warmup uint) (*Handler, error) {
+	r, err := regexp.Compile(p)
+	if err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+	zw, err := zlib.NewWriterLevel(&buf, zlib.BestCompression)
+	if err != nil {
+		return nil, err
+	}
+	writer := compressedWriter{
+		buf:     &buf,
+		zwriter: zw,
+	}
+	scannerBuf := make([]byte, maxScanTokenSize)
+
+	h := Handler{
+		transport:    t,
+		warmupEvents: warmup,
+	}
+	err = fs.WalkDir(storage, ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() || !r.MatchString(path) {
+			return nil
+		}
+
+		f, err := storage.Open(path)
+		if err != nil {
+			return err
+		}
+		s := bufio.NewScanner(f)
+		s.Buffer(scannerBuf, 0)
+		var wroteMeta bool
+		var scanned uint
+		for s.Scan() {
+			line := s.Bytes()
+			if len(line) == 0 {
+				continue
+			}
+
+			if !wroteMeta {
+				writer.WriteLine(line)
+				wroteMeta = true
+				continue
+			}
+			// TODO(marclop): RUM headers should be handled differently.
+			isMeta := bytes.HasPrefix(line, metaHeader) ||
+				bytes.HasPrefix(line, rumMetaHeader)
+			if !isMeta {
+				writer.WriteLine(line)
+				scanned++
+				continue
+			}
+			// Since the current token is a metadata line, it means that we've
+			// read the start of the next batch, because of that, we'll flush,
+			// close and copy the compressed contents. After we've ensured that
+			// we've written the whole batch to the buffer, we reset it and the
+			// zlib.Writer so it can be used for the next iteration.
+			if err := writer.FlushClose(); err != nil {
+				return err
+			}
+			h.batches = append(h.batches, batch{
+				r:     bytes.NewReader(writer.Bytes()),
+				items: scanned,
+			})
+			writer.Reset()
+			writer.WriteLine(line)
+			scanned = 0
+		}
+
+		if err := writer.FlushClose(); err != nil {
+			return err
+		}
+		h.batches = append(h.batches, batch{
+			r:     bytes.NewReader(writer.Bytes()),
+			items: scanned,
+		})
+		buf.Reset()
+		writer.Reset()
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &h, nil
+}
+
+// SendBatches sends the loaded trace data to the configured transport. Returns
+// the total number of documents sent and any transport errors.
+func (h *Handler) SendBatches(ctx context.Context) (uint, error) {
+	var sentEvents uint
+	sendEvents := func(r io.ReadSeeker, events uint) error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			defer r.Seek(0, io.SeekStart)
+		}
+		// TODO(marclop) by reusing the Go agent's transport, we're not able to
+		// set the user agent accordingly. We will need to identify when we have
+		// RUM data and use a different method for sending the traces.
```

Comment:
+1

Reply:
I haven't explored much with regards to this approach. I think that we'll need to implement a different `transport.Transport` which sends the traces to the right URI (`intake/v3/rum/events`)

Reply:
I've added error handling when RUM events are handled on apmbench, that way we can remove the error when we explicitly handling these.